src/lib/core/CHIPError.h

```diff
@@ -114,84 +114,86 @@ class ChipError
+#elif CHIP_CONFIG_ERROR_SOURCE
+    constexpr ChipError(Range range, ValueType value, const char * file = nullptr, unsigned int line = 0) :
+        mError(MakeInteger(range, (value & MakeMask(0, kValueLength)))), mLine(line), mFile(file)
     {}
-#endif // CHIP_CONFIG_ERROR_SOURCE && __cplusplus >= 202002L
+#else
+    constexpr ChipError(Range range, ValueType value) : mError(MakeInteger(range, (value & MakeMask(0, kValueLength)))) {}
+#endif
+
+    /**
+     *  Helper macro to construct a CHIP_ERROR from a range and a value.
+     */
+#define CHIP_GENERIC_ERROR(range, value) ::chip::ChipError(range, value CHIP_ERROR_SOURCE_LOCATION)
 
     /**
      * Construct a CHIP_ERROR for SdkPart @a part with @a code.
      *
      * @note
      *  The macro version CHIP_SDK_ERROR checks that the numeric value is constant and well-formed.
      */
-    constexpr ChipError(SdkPart part, uint8_t code) : ChipError(part, code, /*file=*/nullptr, /*line=*/0) {}
 #if CHIP_CONFIG_ERROR_SOURCE && __cplusplus >= 202002L
-    constexpr ChipError(SdkPart part, uint8_t code, const char * file, unsigned int line,
-                        std::source_location location = std::source_location::current()) :
-        mError(MakeInteger(part, code)) CHIP_INITIALIZE_ERROR_SOURCE(file, line, location)
+    constexpr ChipError(SdkPart part, uint8_t code, std::source_location location = std::source_location()) :
+        mError(MakeInteger(part, code)), mSourceLocation(location)
     {}
-#else
-    constexpr ChipError(SdkPart part, uint8_t code, const char * file, unsigned int line) :
-        mError(MakeInteger(part, code)) CHIP_INITIALIZE_ERROR_SOURCE(file, line, /*loc=*/nullptr)
+#elif CHIP_CONFIG_ERROR_SOURCE
+    constexpr ChipError(SdkPart part, uint8_t code, const char * file = nullptr, unsigned int line = 0) :
+        mError(MakeInteger(part, code)), mLine(line), mFile(file)
     {}
-#endif // CHIP_CONFIG_ERROR_SOURCE && __cplusplus >= 202002L
+#else
+    constexpr ChipError(SdkPart part, uint8_t code) : mError(MakeInteger(part, code)) {}
+#endif
 
     /**
      * Construct a CHIP_ERROR constant for SdkPart @a part with @a code at the current source line.
      * This checks that the numeric value is constant and well-formed.
      * (In C++20 this could be replaced by a consteval constructor.)
      */
-#if CHIP_CONFIG_ERROR_SOURCE
 #define CHIP_SDK_ERROR(part, code)                                                                                                 \
-    (::chip::ChipError(::chip::ChipError::SdkErrorConstant<(part), (code)>::value, __FILE__, __LINE__))
-#else // CHIP_CONFIG_ERROR_SOURCE
-#define CHIP_SDK_ERROR(part, code) (::chip::ChipError(::chip::ChipError::SdkErrorConstant<(part), (code)>::value))
-#endif // CHIP_CONFIG_ERROR_SOURCE
+    (::chip::ChipError(::chip::ChipError::SdkErrorConstant<(part), (code)>::value CHIP_ERROR_SOURCE_LOCATION))
```

Comment:
Why construct `ChipError` using the `SdkErrorConstant` helper and not use the dedicated constructor `constexpr ChipError(SdkPart part, uint8_t code` ?

I suspect this is because of the extra checks with `static_assert` in the `SdkErrorConstant` struct. This however begs the question why do we need the "specialised" `ChipError(SdkPart part, uint8_t code` constructor.